### PR TITLE
Add spambug model

### PR DESCRIPTION
**Pull request by ayush-1506**
This PR still needs revisions.

### CODE REVIEW

**Comment by ayush-1506** [comment]
@marco-c I'm try to do a dryrun (`python -m auto_nag.scripts.spambug --dryrun`) for just one bug, but I still get `Exception: Couldn't get 1 bug classification in 100 seconds, aborting`.

**Comment by marco-c** [comment]
> @marco-c I'm try to do a dryrun (python -m auto_nag.scripts.spambug --dryrun) for just one bug, but I still get Exception: Couldn't get 1 bug classification in 100 seconds, aborting. What's the reason behind this?

Even if you try again? What if you use a larger number, e.g. 10 bugs?

File: auto_nag/scripts/configs/tools.json [hunk 1]
```diff
@@ -565,6 +565,29 @@
...
+        ],
+        "days_lookup": 1,
+        "reporter_skiplist":
+        [
+            "intermittent-bug-filer@mozilla.bugs",
+            "wptsync@mozilla.bugs"
+        ],
+        "confidence_threshold": 0.9,
+        "cc":
+        [
```
Comment (marco-c) [inline auto_nag/scripts/configs/tools.json]: Use an empty list for "cc"

**Comment by ayush-1506** [comment]
@marco-c I tried that but the result is always the same.

File: auto_nag/scripts/configs/tools.json [hunk 2]
```diff
@@ -565,6 +565,29 @@
...
         ]
     },
+    "spambug":
+    {
+        "max_days_in_cache": 1,
+        "receivers":
+        [
+            "calixte@mozilla.com",
+            "sylvestre@mozilla.com",
+            "mcastelluccio@mozilla.com"
```
Comment (marco-c) [inline auto_nag/scripts/configs/tools.json]: Add just `mcastelluccio@mozilla.com` and `ehumphries@mozilla.com`

File: auto_nag/scripts/spambug.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+                continue
+
+            if not {"prob", "index"}.issubset(bug_data.keys()):
+                raise Exception(f"Invalid bug response {bug_id}: {bug_data!r}")
+
+            bug = raw_bugs[bug_id]
+            prob = bug_data["prob"]
+            index = bug_data["index"]
+
+            results[bug_id] = {
```
Comment (marco-c) [inline auto_nag/scripts/spambug.py]: Only add results which have a confidence to be spam higher than the confidence threshold set in the config (`self.get_config("confidence_threshold")`)

**Comment by marco-c** [comment]
> @marco-c I tried that but the result is always the same.

When we talked on Matrix you mentioned you managed to get results with 200 bugs, no?

**Comment by ayush-1506** [comment]
@marco-c Yes, that was for the stepstoreproduce model, not the spambug model.

**Comment by marco-c** [comment]
It seems to be working for me:
```
marco@marco-Dell-XPS ~/Documenti/workspace/relman-auto-nag (master)$ python3
Python 3.7.6 (default, Jan 19 2020, 22:34:52) 
[GCC 9.2.1 20200117] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from auto_nag import bugbug_utils
>>> bugbug_utils.get_bug_ids_classification("spambug", [1608441])
{'1608441': {'class': 0, 'extra_data': {}, 'index': 0, 'prob': [1.0, 0.0]}}
>>> 
```

File: auto_nag/scripts/spambug.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+        return lambda p: (-p[2], -int(p[0]))
+
+    def get_bz_params(self, date):
+        start_date, end_date = self.get_dates(date)
+
+        return {
+            "include_fields": ["id", "groups", "summary"],
+            # Ignore closed bugs.
+            "bug_status": "__open__",
+        }
```
Comment (marco-c) [inline auto_nag/scripts/spambug.py]: We can reduce the number of bugs to retrieve by checking if the reporter doesn't have a mozilla address (as they are surely not spam).

@calixteman can you think of any other bugs to skip?

Comment (marco-c) [reply auto_nag/scripts/spambug.py]: > We can reduce the number of bugs to retrieve by checking if the reporter doesn't have a mozilla address (as they are surely not spam).

Adding `f1=reporter&v1=@mozilla&o1=notsubstring` should take care of it.

Comment (calixteman) [reply auto_nag/scripts/spambug.py]: probably we can filter after the query ran in using `people::People::is_mozilla`: some people may be mozilla but with a different bugmail.
Anyway, I don't see any other pattern we could filter out.

File: templates/spambug.html [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<p>The following {{ plural('bug', data, pword='bugs') }} {{ plural('doesn\'t', data, pword='don\'t') }} are probably spam :
```
Comment (calixteman) [inline templates/spambug.html]: nano-nit: `are` should be `is` when data contains only one bug

File: auto_nag/scripts/configs/tools.json [hunk 6]
```diff
@@ -565,6 +565,23 @@
             "jdescottes@mozilla.com"
         ]
     },
+    "spambug":
+    {
+        "max_days_in_cache": 1,
+        "receivers":
+        [
+            "ehumphries@mozilla.com",
+            "mcastelluccio@mozilla.com"
```
Comment (marco-c) [inline auto_nag/scripts/configs/tools.json]: Actually, could you add sylvestre@mozilla.com back? Sylvestre was interested in getting the email too.

**Comment by marco-c** [comment]
We also need to add the script to https://github.com/mozilla/relman-auto-nag/blob/fd1011b16ed84d4887ea0b8041165152dded1d57/runauto_nag_daily.sh, so that it runs daily.

File: templates/spambug.html (hunk 5 above)
Comment (marco-c) [reply templates/spambug.html]: And `{{ plural('doesn\'t', data, pword='don\'t') }}` should be removed I guess

File: auto_nag/scripts/spambug.py (hunk 4 above)
Comment (marco-c) [reply auto_nag/scripts/spambug.py]: OK, let's do both:
- if we just filter out after the query, we spend time classifying some bugs needlessly;
- if we just filter before the query, we will miss some people who are actually mozillians.

File: auto_nag/scripts/spambug.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+        return "[Using ML] Detect spam bugs"
+
+    def columns(self):
+        return ["id", "summary", "confidence", "autofixed"]
+
+    def sort_columns(self):
+        return lambda p: (-p[2], -int(p[0]))
+
+    def handle_bug(self, bug, data):
+        reporter = bug["creator"]
```
Comment (marco-c) [inline auto_nag/scripts/spambug.py]: You're using `creator`, but `creator` is not in the include_fields in the Bugzilla query, so it won't be available. You'll need to add it in the include_fields list.

File: auto_nag/scripts/configs/tools.json [hunk 8]
```diff
@@ -565,6 +565,24 @@
             "jdescottes@mozilla.com"
         ]
     },
+    "spambug":
+    {
+        "max_days_in_cache": 1,
```
Comment (marco-c) [inline auto_nag/scripts/configs/tools.json]: Use 7 for the cache, as the results are not going to change

File: auto_nag/scripts/configs/tools.json [hunk 9]
```diff
@@ -565,6 +565,24 @@
...
+        "receivers":
+        [
+            "ehumphries@mozilla.com",
+            "mcastelluccio@mozilla.com",
+            "sylvestre@mozilla.com"
+        ],
+        "days_lookup": 1,
```
Comment (marco-c) [inline auto_nag/scripts/configs/tools.json]: Use 3 days here, like for stepstoreproduce and other bugbug-based tools

File: auto_nag/scripts/configs/tools.json [hunk 10]
```diff
@@ -565,6 +565,24 @@
...
+    {
+        "max_days_in_cache": 7,
+        "receivers":
+        [
+            "ehumphries@mozilla.com",
+            "mcastelluccio@mozilla.com",
+            "sylvestre@mozilla.com"
+        ],
+        "days_lookup": 3,
+        "reporter_skiplist":
```
Comment (marco-c) [inline auto_nag/scripts/configs/tools.json]: You can remove this, as it is not being used

File: auto_nag/scripts/spambug.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+        start_date, end_date = self.get_dates(date)
+
+        return {
+            "include_fields": ["id", "groups", "summary", "creator"],
+            # Ignore closed bugs.
+            "bug_status": "__open__",
+            "f1": "reporter",
+            "v1": "@mozilla",
+            "o1": "notsubstring",
+        }
```
Comment (marco-c) [inline auto_nag/scripts/spambug.py]: A couple more ideas to reduce the set of bugs, then we should be good to go.

- Skip bugs reported by Softvision (QA):
```
            "f2": "reporter",
            "v2": "@softvision",
            "o2": "notsubstring",
```

- Actually make use of start_date, considering only bugs which were recently opened:
```
            "f3": "creation_ts",
            "o3": "greaterthan",
            "v3": start_date,
```

Comment (calixteman) [reply auto_nag/scripts/spambug.py]: +1

File: auto_nag/scripts/spambug.py [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+                continue
+
+            if not {"prob", "index"}.issubset(bug_data.keys()):
+                raise Exception(f"Invalid bug response {bug_id}: {bug_data!r}")
+
+            bug = raw_bugs[bug_id]
+            prob = bug_data["prob"]
+            index = bug_data["index"]
+
+            if prob[index] > self.get_config("confidence_threshold"):
```
Comment (marco-c) [inline auto_nag/scripts/spambug.py]: We only want to show bugs which we detected as spam, so only if index == 1.
You can simply do `if prob[1] > self.get_config("confidence_threshold"):`.

**Comment by marco-c** [comment]
Thanks @ayush-1506!

**Pull request merged**